src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/services/HdfsService.scala

```diff
@@ -0,0 +1,167 @@
+  private lazy val conf = SparkHadoopUtil.get.conf
+  private lazy val fs = FileSystem.get(conf)
+
+  /**
+   *  See org.apache.spark.sql.execution.streaming.HDFSMetadataLog
+   */
+  private val batchFilesFilter = new PathFilter {
+    override def accept(path: Path): Boolean = {
+      try {
+        path.getName.toLong
+        true
+      } catch {
+        case _: NumberFormatException =>
+          false
+      }
+    }
+  }
+
+  /**
+   *  @param pathStr path to the file as a string
+   *  @param parseFn function that parses the file line by line. Caution: It must materialize the content,
+   *                because the file is closed after the method completes. E.g. it must not return an iterator.
+   *  @tparam R type of the parsed value
+   *  @return None if the file doesn't exist, Some with the parsed content
+   */
+  override def parseFileAndClose[R](pathStr: String, parseFn: Iterator[String] => R): Option[R] = {
+    val path = new Path(pathStr)
+    if (fs.exists(path)) {
+      val input = fs.open(path)
+      try {
+        val lines = Source.fromInputStream(input, UTF_8.name()).getLines()
+        Some(parseFn(lines))
+      } catch {
+        case e: Exception =>
+          // re-throw the exception with the log file path added
+          throw new Exception(s"Failed to parse file $path", e)
+      } finally {
+        IOUtils.closeQuietly(input)
+      }
+    } else {
+      logger.debug(s"Could not find file $path")
+      None
+    }
+  }
+
+  /**
+   *  see org.apache.spark.sql.execution.streaming.OffsetSeqLog
+   *  and org.apache.spark.sql.kafka010.JsonUtils
+   *  for details on the assumed format
+   */
+  override def parseKafkaOffsetStream(lines: Iterator[String]): TopicPartitionOffsets = {
+    val SERIALIZED_VOID_OFFSET = "-"
+    def parseOffset(value: String): Option[TopicPartitionOffsets] = value match {
+      case SERIALIZED_VOID_OFFSET => None
+      case json                   => Some(mapper.readValue(json, classOf[TopicPartitionOffsets]))
+    }
+    if (!lines.hasNext) {
+      throw new IllegalStateException("Incomplete log file")
+    }
+
+    lines.next() // skip version
+    lines.next() // skip metadata
+    lines
+      .map(parseOffset)
+      .filter(_.isDefined)
+      .map(_.get)
+      .toSeq
+      .head
+  }
+
+  /**
+   *  @return an Option of a String, Boolean pair. The string contains the path to the latest offset file, while the
+   *         boolean is true if the offset is committed (i.e. a corresponding commit file exists), and false otherwise.
+   *         None is returned if the offset file does not exist. If the offset file does not exist, the corresponding
+   *         commit file is assumed to also not exist.
+   */
+  override def getLatestOffsetFilePath(params: HdfsParameters): Option[(String, Boolean)] = {
+    userGroupInformationWrapper.loginUserFromKeytab(params.principal, params.keytab)
+    val offsetBatchIdOpt = getLatestOffsetBatchId(params.checkpointLocation)
```

Comment:
Is this a good logic ? If I do understand correctly. 
You get latest offset and then you try to find commit for it. 
Why don't you find latest commit and based on that get an offset file ?

Reply:
If the latest offset file doesn't exist, then there's no point in searching for the latest commit file. I don't know what will be more frequent